README.md
hateoas-client.js README
=======================

* Latest Release: [![GitHub version](https://badge.fury.io/gh/DracoBlue%2Fhateoas-client-js.png)](https://github.com/DracoBlue/hateoas-client-js/releases)
* Build-Status: [![Build Status](https://travis-ci.org/DracoBlue/hateoas-client-js.png?branch=master)](https://travis-ci.org/DracoBlue/hateoas-client-js)
* Official Site: <http://dracoblue.net/>

hateoas-client.js is copyright 2011-2015 by DracoBlue <http://dracoblue.net>

What is hateoas-client.js?
-----------------------

hateoas-client.js is a library (for browser+nodejs) to communicate with RESTful services. It uses
jQuery as ajax library. It's aim is to provide a very simple API to follow
the `links` defined in a request response, thus achieving
level 3 in `Richardson Maturity Model`.

Requirements:

* jQuery 1.5+

Installation
------------

* On the browser: `$ bower install hateoas-client`
* In nodejs: `$ npm install hateoas-client`

How does it work?
-----------------

### Example with JSON (in the browser)

If you include `hateoas-client.js` after your `jQuery.js`, you'll have the ability
to make such requests:

    var a = new HttpAgent('/api');
    a.navigate(['coffee', {'name': 'Small'}, "buy"]);
    a.post(function(response) {
        // response.getValue() contains what the POST /api/orders?product_id=5 returned
    });

This example assumes that the responses contain

    GET /api
        {
            "links": [ { "rel": "coffee", "href": "/api/coffee" } ]
        }
    GET /api/coffee
        [
            {
                "id": 5,
                "name": "Small",
                "links": [ { "rel": "buy", "href": "/api/orders?product_id=5" } ]
            }
        ]

### Remote Example with Atom-Feed

This example retrieves the most viewed videos from youtube, navigates 2x next, chooses the
first element (because of empty filter`{}`), navigates to it's `self` link and finally:
returns the `<title>` element's text.

    var a = new HttpAgent('http://gdata.youtube.com/feeds/api/standardfeeds/most_viewed?max-results=5', {}, {
        'proxy_script': 'proxy.php?url='
    });

    a.navigate(['next', 'next', {}, 'self']);
    a.get(function(response) {
        var title = jQuery(response.getValue()).find('title').text();
    });            

That example also shows, how one can use the `proxy_script`-option to use a
`.php`-script to retrieve contents from a remote site.

Usage with require.js
---------------------

If you want to retrieve the HttpAgent in your require.js script use (ensure that `hateoas-client` maps on `hateoas-client.js`
in your requirejs config file):

``` javascript
require('hateoas-client', function(hateoasClient) {
    var a = new hateoasClient.HttpAgent('/api');
});
```

Supported Media Types
---------------------

* `application/json` (detected by `JsonHttpResponse`), supported features:
	- `getValue()`: Returns the entire response as json object
	- `getValues()`: If the entire response was an array, it will return each element as `JsonHttpResponse[]`.
	- `getMatchingValue()`: Filters on `getValues()`
	- `getLinks()`: Expects a `links`-property as array with `{rel: REL, "href": URL}` elements and returns them as `HttpLink[]`
* `application/atom+xml` (detected by `AtomXmlHttpResponse`), supported features:
	- `getValue()`: Returns the entire response as xml object
	- `getValues()`: Returns all `<entry>` children as `AtomXmlHttpResponse[]`.
	- `getMatchingValue()`: Filters on `getValues()`
	- `getLinks()`: Returns all `<link>` children as `HttpLink[]`
* `application/hal+json` (detected by `JsonHalHttpResponse`), supported features:
	- `getValue()`: Returns the entire response as json object
	- `getValues()`: Returns all `_embedded` objects as `JsonHalHttpResponse[]`.
	- `getMatchingValue()`: Filters on `getValues()`
	- `getLinks()`: Returns all `_links` and `_embedded` links as `HttpLink[]` according to HAL specification
* `application/hc+json` (detected by `JsonHcHttpResponse`), supported features:
	- `getValue()`: Returns the entire response as json object
	- `getValues()`: Returns all embedded objects on root level with `self`-link as `JsonHalHttpResponse[]`.
	- `getMatchingValue()`: Filters on `getValues()`
	- `getLinks()`: Returns all embedded objects (with `self`-link), obvious links (with iana registration) or properties starting with `http:`/`https:` on root level as `HttpLink[]`
* `application/xml` (detected by `XmlHttpResponse`), supported features:
	- `getValue()`: Returns the entire response as xml object

Todos
-----

* test and extend support for other responses (xml, maybe a generic converter system or usage of the one from jQuery)
* handle status codes other then 200 (currently only 200 is `JsonHttpResponse#isOk() == true` and 201 is interpreted)
* add documentation for `HttpAgent`, `JsonHttpResponse`, `AtomXmlHttpResponse` and `XmlHttpResponse`
* ... more as soon as I get to that!

Changelog
---------

See CHANGLOG.md for more information.

License
--------

hateoas-client.js is licensed under the terms of MIT. See LICENSE for more information.


CHANGELOG.md
hateoas-client.js Changelog
=======================

## dev

  - added experimental support for JSON HC
  - added tests for hal in `test/hal-test.js`
  
## 0.4.0 (2015/04/21)

  - added `BaseHttpResponse#getAllHeaders`
  - added `BaseHttpResponse#getStatusCode`
  - added `HttpAgent#getUrl`
  - added `HttpLink` as subclass of HttpAgent with `getRel`, `getType` and `getTitle`
  - added `head` request
  - added `NoContentResponse`
  - added `options.ajaxOptions` to override `jQuery.ajax` options

## 0.3.1 (2015/04/20)

  - fixed multiple links in hal

## 0.3.0 (2015/04/06)

  - in case of unsupported media type return `UnsupportedMediaTypeHttpResponse`
  - added server test
  - inject agent also in subresponses of json, hal and atom response
  - handle `Location`-redirects with `/` at the beginning on 201 Status Code
  - added `HttpAgent#logDebug` and `HttpAgent#logTrace` (enable it with `HttpAgent.enableLogging=true;` )

## 0.2.0 (2015/04/04)

  - added mocha test infrastructure
  - renamed to hateoas-client.js
  - handle relative paths in links (by asking HttpAgent for the base url)
  - Added FIXME method for getLinks on HTML/XML objects
  - added nodejs support with domino, jquery and xmlhttprequest for nodejs
  - added `HttpAgent#getBaseUrl`

## 0.1.0

  - added definition for requirejs
  - added JsonHalHttpResponse for HAL hyper media type
  - fixes JsonHttpResponse xhr variable on sub values
  - added AtomXmlHttpResponse and example for atom feed retrieval
  - added filtering * + object to breadth first search for anything until it matches a filter object
  - added handler for 201 Created response
  - added handling for status code 200
  - added * as indicator for breadth first search in HttpAgent#navigate
  - added function as filter object
  - example files added
  - initial commit

test/hal-test.js
var assert = require("assert")
var hateoasClient = require('./../hateoas-client.js');

describe('HAL Test', function(){
	var express = require('express');
	var app = express();

	//hateoasClient.HttpAgent.enableLogging = true;

	app.get('/api.json', function(req, res) {
		res.set('Content-Type', 'application/hal+json');
		res.send(JSON.stringify({
			"_links": { "coffee": { "href": "/api/coffee.json" } }
		}));
	});
    app.get('/api/product/5', function(req, res) {
        res.set('Content-Type', 'application/hal+json');
        res.send(JSON.stringify(
            {
                "id": 5,
                "name": "Small",
                "_links": {"self": {"href": "/api/product/5"}, "buy": {"href": "/api/submitted-orders?product_id=5"}}
            }
        ));
    });
	app.get('/api/coffee.json', function(req, res) {
		res.set('Content-Type', 'application/hal+json');
		res.send(JSON.stringify({
            "title": "Coffee Menu",
            "_links": {
                "product": [{"title": "Small", "href": "/api/product/5"}]
            },
            "_embedded": {
                "product": [
                    {
                        "id": 5,
                        "name": "Small",
                        "_links": {"self": {"href": "/api/product/5"}, "buy": {"href": "/api/submitted-orders?product_id=5"}}
                    }
                ]
            }
        }));
	});
	app.post('/api/submitted-orders', function(req, res) {
		res.redirect(201, '/api/orders/1');
	});
	app.get('/api/orders/1', function(req, res) {
		res.set('Content-Type', 'application/hal+json');
		res.send(JSON.stringify(
			{
				"id": 1,
				"status": "pending"
			}
		));
	});

	describe('new HttpAgent()', function(){

		var server = null;

		before(function() {
			server = app.listen(3000, '127.0.0.1');
		});

		after(function(done){
			server.close(function() {
				done();
			});
		});

        it('should give a link', function(done){
            var agent = new hateoasClient.HttpAgent('http://127.0.0.1:3000/api.json');

            agent.get(function(response) {
                assert.equal(true, response.isOk(), 'It was not possible to navigate!');
                var links = response.getLinks();
                assert.ok(links["coffee"]);
                assert.equal(1, links["coffee"].length);
                var coffee_link = response.getLink('coffee');
                assert.equal('http://127.0.0.1:3000/api/coffee.json', coffee_link.getUrl());
                assert.equal('coffee', coffee_link.getRel());
                assert.equal('Express', response.getHeader('x-powered-by'));
                assert.equal('Express', response.getAllHeaders()['x-powered-by']);
                done();
            });
        });

		it('should navigate with filter object', function(done){
			var agent = new hateoasClient.HttpAgent('http://127.0.0.1:3000/api.json');

			agent.navigate(['coffee', {
				"name" : 'Small'
			}, 'buy']);

			agent.post(function(response) {
				assert.equal(true, response.isOk(), 'It was not possible to navigate!');
				assert.equal('pending', response.getValue().status, 'Cannot find status after navigation');
				done();
			});
		});

		it('should navigate without anything', function(done){
			var agent = new hateoasClient.HttpAgent('http://127.0.0.1:3000/api.json');

			agent.navigate('coffee');

			agent.get(function(response) {
				assert.equal(true, response.isOk(), 'It was not possible to navigate!');
				assert.equal('Coffee Menu', response.getValue().title, 'Cannot find title of the coffee menu');
				done();
			});
		});

		it('should navigate with filter function', function(done){
			var agent = new hateoasClient.HttpAgent('http://127.0.0.1:3000/api.json');

			agent.navigate(['coffee', function(value) {
				return value.name === 'Small';
			}, 'buy']);

			agent.post(function(response) {
				assert.equal(true, response.isOk(), 'It was not possible to navigate!');
				assert.equal('pending', response.getValue().status, 'Cannot find status after navigation');
				done();
			});
		});

		it('should return isOk = false, in case of broken link', function(done){
			var agent = new hateoasClient.HttpAgent('http://127.0.0.1:3000/api.json');

			agent.navigate(['coffee', 'invalid_link', 'buy']);

			agent.get(function(response) {
				assert.equal(false, response.isOk(), 'It was possible to navigate!');
				done();
			});
		});
	})
})

test/init-test.js
var assert = require("assert")
var hateoasClient = require('./../hateoas-client.js');

describe('InitializeHttpAgent', function(){
	describe('new HttpAgent()', function(){

		it('should initialize an agent', function(){
			var agent = new hateoasClient.HttpAgent();
		})
	})
})

test/server-test.js
var assert = require("assert")
var hateoasClient = require('./../hateoas-client.js');

describe('ServerTest', function(){
	var express = require('express');
	var app = express();

	//hateoasClient.HttpAgent.enableLogging = true;

	app.get('/api.json', function(req, res) {
		res.set('Content-Type', 'application/json');
		res.send(JSON.stringify({
			"links": [ { "rel": "coffee", "href": "/api/coffee.json" }]
		}));
	});
	app.get('/api/coffee.json', function(req, res) {
		res.set('Content-Type', 'application/json');
		res.send(JSON.stringify([
			{
				"id": 5,
				"name": "Small",
				"links": [ { "rel": "buy", "href": "/api/submitted-orders?product_id=5" } ]
			}
		]));
	});
	app.post('/api/submitted-orders', function(req, res) {
		res.redirect(201, '/api/orders/1');
	});
	app.get('/api/orders/1', function(req, res) {
		res.set('Content-Type', 'application/json');
		res.send(JSON.stringify(
			{
				"id": 1,
				"status": "pending"
			}
		));
	});

	describe('new HttpAgent()', function(){

		var server = null;

		before(function() {
			server = app.listen(3000, '127.0.0.1');
		});

		after(function(done){
			server.close(function() {
				done();
			});
		});

        it('should give a link', function(done){
            var agent = new hateoasClient.HttpAgent('http://127.0.0.1:3000/api.json');

            agent.get(function(response) {
                assert.equal(true, response.isOk(), 'It was not possible to navigate!');
                var links = response.getLinks();
                assert.ok(links["coffee"]);
                assert.equal(1, links["coffee"].length);
                var coffee_link = response.getLink('coffee');
                assert.equal('http://127.0.0.1:3000/api/coffee.json', coffee_link.getUrl());
                assert.equal('coffee', coffee_link.getRel());
                assert.equal('Express', response.getHeader('x-powered-by'));
                assert.equal('Express', response.getAllHeaders()['x-powered-by']);
                done();
            });
        });

		it('should navigate with filter object', function(done){
			var agent = new hateoasClient.HttpAgent('http://127.0.0.1:3000/api.json');

			agent.navigate(['coffee', {
				"name" : 'Small'
			}, 'buy']);

			agent.post(function(response) {
				assert.equal(true, response.isOk(), 'It was not possible to navigate!');
				assert.equal('pending', response.getValue().status, 'Cannot find status after navigation');
				done();
			});
		});

		it('should navigate without anything', function(done){
			var agent = new hateoasClient.HttpAgent('http://127.0.0.1:3000/api.json');

			agent.navigate('coffee');

			agent.get(function(response) {
				assert.equal(true, response.isOk(), 'It was not possible to navigate!');
				assert.equal('Small', response.getValue()[0].name, 'Cannot find name of first coffee');
				done();
			});
		});

		it('should navigate with filter function', function(done){
			var agent = new hateoasClient.HttpAgent('http://127.0.0.1:3000/api.json');

			agent.navigate(['coffee', function(value) {
				return value.name === 'Small';
			}, 'buy']);

			agent.post(function(response) {
				assert.equal(true, response.isOk(), 'It was not possible to navigate!');
				assert.equal('pending', response.getValue().status, 'Cannot find status after navigation');
				done();
			});
		});

		it('should return isOk = false, in case of broken link', function(done){
			var agent = new hateoasClient.HttpAgent('http://127.0.0.1:3000/api.json');

			agent.navigate(['coffee', 'invalid_link', 'buy']);

			agent.get(function(response) {
				assert.equal(false, response.isOk(), 'It was possible to navigate!');
				done();
			});
		});
	})
})

rest-client.js
/*
 * Copyright (c) 2011-2015 DracoBlue, http://dracoblue.net/
 *
 * Licensed under the terms of MIT License. For the full copyright and license
 * information, please see the LICENSE file in the root folder.
 */

if (typeof window === 'undefined') { /* Running in NodeJS */
	var domino = require('domino');
	var $ = require('jquery')(domino.createWindow());
	var jQuery = $;
	var XMLHttpRequest = require('xmlhttprequest').XMLHttpRequest;
	$.support.cors=true; // cross domain
	$.ajaxSettings.xhr = function() {
		return new XMLHttpRequest();
	};
}

NotOkHttpResponse = function() {
    
};

NotOkHttpResponse.prototype.isOk = function() {
    return false;
};

HttpAgent = function(url, headers, options) {
    this.url = url;
    this.default_headers = headers || {};
    this.navigation_steps = [];
    this.options = options || {};
};

HttpAgent.prototype.clone = function() {
    var clone = new HttpAgent(this.url);
    clone.default_headers = jQuery.extend(true, {}, this.default_headers);
    clone.navigation_steps = jQuery.extend(true, [], this.navigation_steps);
    return clone;
};

HttpAgent.prototype.getBaseUrl = function()
{
	var match = this.url.match(/^(.+?[\/]+?.+?)\//);
	if (match)
	{
		return match[1];
	}

	return '';
};

HttpAgent.prototype.rawCall = function(cb, verb, params, headers) {
    var that = this;
    
    headers = headers || {};
    var url = that.url;
    
    if (this.options.proxy_script) {
        url = this.options.proxy_script + encodeURIComponent(url);
    }
    
    jQuery.ajax({
        beforeSend: function(xhrObj){
            for (var header in that.default_headers)
                xhrObj.setRequestHeader(header, that.default_headers[header]);
            for (var header in headers)
                xhrObj.setRequestHeader(header, headers[header]);
            return xhrObj;
        },        
        url: url,
        dataType: 'text',
        type: verb,
        data: params || {},
        complete: function(response) {
            if (response.status === 201) {
                that.url = response.getResponseHeader('Location');
                that.rawCall(cb, 'GET', {}, headers);
            } else {
                cb(HttpAgent.getHttpResponseByRawResponse(response, that));
            }
        }
    });
};

HttpAgent.prototype.rawNavigate = function(cb) {
    var that = this;
    var step_position = 0;
    var last_response = null;
    
    var performNextStep = function() {
        if (step_position === that.navigation_steps.length) {
            that.navigation_steps = [];
            cb(last_response);
            return ;
        }
        
        that.rawCall(function(current_response) {
            last_response = current_response;
            if (!current_response.isOk()) {
                cb(current_response);
                return ;
            }
            
            var next_step = that.navigation_steps[step_position];
            
            if (typeof next_step !== 'string') {
                step_position += 1;
                current_response = current_response.getMatchingValue(next_step);
            }
            
            var link_name = that.navigation_steps[step_position];
            
            if (link_name === '*') {
                step_position++;
                var filter_object = that.navigation_steps[step_position];
                
                that.rawBreadthFirstSearch(function(next_step_entry_point, last_search_response) {
                    last_response = last_search_response;
                    if (!next_step_entry_point) {
                        cb(new NotOkHttpResponse());
                        return ;
                    }
                    that.url = next_step_entry_point.url;
                    step_position++;
                    performNextStep();
                }, filter_object);
            } else {
                var next_step_entry_point = null;
                try {
                    next_step_entry_point = current_response.getLink(link_name);
                } catch (error) {
                    cb(new NotOkHttpResponse());
                    return;
                }
                that.url = next_step_entry_point.url;
                
                step_position++;
                performNextStep();
            }
        }, 'GET');
    };
    
    performNextStep();
};

HttpAgent.prototype.rawBreadthFirstSearch = function(cb, filter_object) {
    var url_was_in_frontier = {};
    url_was_in_frontier[this.url] = true;
    
    var frontier = [this.url];
    var frontier_length = frontier.length;
    var tmp_entry_point = new HttpAgent(this.url);

    var new_frontier = [];
    
    var step_position = 0;
    
    var performIteration = function() {
        if (step_position === frontier_length && new_frontier.length === 0) {
            cb(null, new NotOkHttpResponse());
            return ;
        }

        if (step_position === frontier_length) {
            step_position = 0;
            frontier = new_frontier;
            frontier_length = frontier.length;
            new_frontier = [];
        }
        
        tmp_entry_point.url = frontier[step_position];
        tmp_entry_point.rawCall(function(response) {
            /*
             * Was this response ok?
             */
            if (response.isOk()) {
                var links = response.getLinks();
                
                if (typeof filter_object === "string" && typeof links[filter_object] !== 'undefined') {
                    /*
                     * YES!
                     */
                    cb(links[filter_object][0], response);
                    return ;
                }
                
                if (typeof filter_object !== "string") {
                    try {
                        var matching_value = response.getMatchingValue(filter_object);
                        cb(tmp_entry_point.clone(), response);
                        return ;
                    } catch (error) {
                        /*
                         * We'll continue, if that object didn't match
                         */
                    }
                }
                
                jQuery.each(links, function(pos, link_targets) {
                    jQuery.each(link_targets, function(sub_pos, link) {
                        if (!url_was_in_frontier[link.url]) {
                            new_frontier.push(link.url);
                            url_was_in_frontier[link.url] = true;
                        }
                    });
                });
            }
            
            /*
             * Continue with next element in the frontier!
             */
            step_position++;
            performIteration();
        }, 'GET');
    };
    
    performIteration();
};


HttpAgent.prototype.call = function(cb, verb, params, headers) {
    var that = this;
    if (this.navigation_steps.length === 0) {
        this.rawCall(cb, verb, params, headers);
    } else {
        this.rawNavigate(function(navigation_response) {
            if (navigation_response.isOk()) {
                that.rawCall(cb, verb, params, headers);
            } else {
                cb(navigation_response);
            }
        });
    }
};

HttpAgent.prototype.get = function(cb, params, headers) {
    this.call(cb, 'GET', params, headers || {});
};

HttpAgent.prototype.post = function(cb, params, headers) {
    this.call(cb, 'POST', params, headers || {});
};

HttpAgent.prototype['delete'] = function(cb, params, headers) {
    this.call(cb, 'DELETE', params, headers || {});
};

HttpAgent.prototype.put = function(cb, params, headers) {
    this.call(cb, 'PUT', params, headers || {});
};

HttpAgent.prototype.patch = function(cb, params, headers) {
    this.call(cb, 'PATCH', params, headers || {});
};

HttpAgent.prototype.navigate = function(steps) {
    if (typeof steps === 'string') {
        this.addNavigationStep(steps);
    } else {
        var steps_length = steps.length;
        for (var i = 0; i < steps_length; i++) {
            this.addNavigationStep(steps[i]);
        }
    }
    return this;
};

HttpAgent.prototype.addNavigationStep = function(step) {
    this.navigation_steps.push(step);
};

HttpAgent.response_content_types = [];

HttpAgent.registerResponseContentTypes = function(content_types, converter_class) {
    this.response_content_types.push([content_types, converter_class]);
};
        
HttpAgent.getHttpResponseByRawResponse = function(raw_response, agent) {
    var content_type = (raw_response.getResponseHeader('content-type') || '').toLowerCase().split(';')[0];
    var response_content_types = this.response_content_types;
    var response_content_types_length = response_content_types.length;
    for (var i = 0; i < response_content_types_length; i++) {
        if (response_content_types[i][0].indexOf(content_type) !== -1) {
            var converter_class = response_content_types[i][1];
            return new converter_class(raw_response, null, agent);
        }
    }
    
    throw new Error('Cannot find converter for content type: ' + content_type);
};

BaseHttpResponse = function() {
    
};

BaseHttpResponse.prototype.isOk = function() {
    return this.xhr.status === 200 ? true : false;
};

BaseHttpResponse.prototype.getHeader = function(name, default_value) {
    return this.xhr.getResponseHeader(name) || default_value;
};

BaseHttpResponse.prototype.getLink = function(link_name) {
    var links = this.getLinks();
    if (typeof links[link_name] === 'undefined') {
        throw new Error('Cannot find link with name: ' + link_name);
    }
    return links[link_name][0];
};

JsonHttpResponse = function(xhr, value, agent) {
    this.xhr = xhr;
	this.agent = agent;
    this.value = value || null;
    this.values = null;
    this.links_map = null;
};

jQuery.extend(JsonHttpResponse.prototype, BaseHttpResponse.prototype);

JsonHttpResponse.prototype.getValue = function() {
    if (!this.value) {
        this.value = jQuery.parseJSON(this.xhr.responseText);
    }
    
    return this.value;
};

JsonHttpResponse.prototype.at = function(pos) {
    var values = this.getValues();
    return values[pos];
};

JsonHttpResponse.prototype.getValues = function() {
    if (!this.values) {
        var value_entry_points = [];

        var values = this.getValue(this.xhr);
        var values_length = values.length;

        for (var i = 0; i < values_length; i++) {
            value_entry_points.push(new JsonHttpResponse(this.xhr, values[i]));
        }
        
        this.values = value_entry_points;
    }
    
    return this.values;
};

JsonHttpResponse.prototype.getMatchingValue = function(filter_object) {
    var value_entry_points = [];
    
    var values = this.getValue();
    
    /*
     * FIXME: An old man's check, whether what we've got here is an array or not
     */
    if (typeof values !== "object" || typeof values.join !== "function") {
        values = [values];
    }
    
    var values_length = values.length;
    
    for (var i = 0; i < values_length; i++) {
        var value = values[i];
        var is_match = true;
        if (typeof filter_object === 'function') {
            is_match = filter_object(value);
        } else {
            for (key in filter_object) {
                if (filter_object.hasOwnProperty(key) && filter_object[key] !== value[key]) {
                    is_match = false;
                }
            }
        }
        
        if (is_match) {
            return new JsonHttpResponse(this.xhr, value);
        }
    }
    
    throw new Error('No matching value found for filter object');
};

JsonHttpResponse.prototype.getLinks = function() {
    if (this.links_map) {
        return this.links_map;
    }
    
    var value = this.getValue();
    var links = [];

    if (value.hasOwnProperty('links')) {
        links = jQuery.extend(true, links, value['links']);
    }
    
    if (value.hasOwnProperty('link')) {
        links.push(value['link']);
    }
    
    var links_map = {};
    var links_length = links.length;
    
    for (var i = 0; i < links_length; i++) {
        var link = links[i];
        var headers = {};
        if (link.type) {
            headers['Content-Type'] = link.type;
        }
        links_map[link.rel] = links_map[link.rel] || [];
		var absolute_href = link.href;
		if (absolute_href.substr(0, 1) == '/')
		{
			absolute_href = this.agent.getBaseUrl() + absolute_href;
		}
        links_map[link.rel].push(new HttpAgent(absolute_href, headers));
    }
    
    this.links_map = links_map;
    return this.links_map;
};

HttpAgent.registerResponseContentTypes(['application/json'], JsonHttpResponse);

JsonHalHttpResponse = function(xhr, value, agent) {
    this.xhr = xhr;
	this.agent = agent;
    this.value = value || null;
    this.values = null;
    this.links_map = null;
};

jQuery.extend(JsonHalHttpResponse.prototype, BaseHttpResponse.prototype);

JsonHalHttpResponse.prototype.getValue = function() {
    if (!this.value) {
        this.value = jQuery.parseJSON(this.xhr.responseText);
    }

    return this.value;
};

JsonHalHttpResponse.prototype.at = function(pos) {
    var values = this.getValues();
    return values[pos];
};

JsonHalHttpResponse.prototype.getValues = function() {
    if (!this.values) {
        var value_entry_points = [];

        var value = this.getValue(this.xhr);

        if (value.hasOwnProperty('_embedded')) {
            embedded_objects_map = jQuery.extend(true, {}, value['_embedded']);
        }

        for (var rel in embedded_objects_map)
        {
            if (embedded_objects_map.hasOwnProperty(rel))
            {
                var embedded_objects = embedded_objects_map[rel];

                /*
                 * FIXME: An old man's check, whether what we've got here is an array or not
                 */
                if (typeof embedded_objects !== "object" || typeof embedded_objects.join !== "function") {
                    embedded_objects = [embedded_objects];
                }

                var embedded_objects_length = embedded_objects.length;

                for (var i = 0; i < embedded_objects_length; i++)
                {
                    value_entry_points.push(new JsonHalHttpResponse(this.xhr, embedded_objects[i]));
                }
            }
        }

        this.values = value_entry_points;
    }

    return this.values;
};

JsonHalHttpResponse.prototype.getMatchingValue = function(filter_object) {
    var value_entry_points = [];

    var values = this.getValues();

    var values_length = values.length;

    for (var i = 0; i < values_length; i++) {
        var value = values[i].getValue();
        var is_match = true;
        if (typeof filter_object === 'function') {
            is_match = filter_object(value);
        } else {
            for (key in filter_object) {
                if (filter_object.hasOwnProperty(key) && filter_object[key] !== value[key]) {
                    is_match = false;
                }
            }
        }

        if (is_match) {
            return new JsonHalHttpResponse(this.xhr, value);
        }
    }

    throw new Error('No matching value found for filter object');
};

JsonHalHttpResponse.prototype.getLinks = function() {
    if (this.links_map) {
        return this.links_map;
    }

    var value = this.getValue();
    var links_map = {};
    var raw_links_map = {};

    if (value.hasOwnProperty('_links')) {
        raw_links_map = jQuery.extend(true, raw_links_map, value['_links']);
    }

    for (var rel in raw_links_map)
    {
        if (raw_links_map.hasOwnProperty(rel))
        {
            var raw_links = raw_links_map[rel];

            /*
             * FIXME: An old man's check, whether what we've got here is an array or not
             */
            if (typeof raw_links !== "object" || typeof raw_links.join !== "function") {
                raw_links = [raw_links];
            }

            var raw_links_length = raw_links.length;

            for (var i = 0; i < raw_links_length; i++)
            {
                var link = raw_links[i];
                var headers = {};
                /* FIXME: is `type` allowed in HAL? */
                if (link.type) {
                    headers['Content-Type'] = link.type;
                }
                links_map[rel] = links_map[link.rel] || [];
				var absolute_href = link.href;
				if (absolute_href.substr(0, 1) == '/')
				{
					absolute_href = this.agent.getBaseUrl() + absolute_href;
				}
                links_map[rel].push(new HttpAgent(absolute_href, headers));
            }
        }
    }

    this.links_map = links_map;
    return this.links_map;
};

HttpAgent.registerResponseContentTypes(['application/hal+json'], JsonHalHttpResponse);


AtomXmlHttpResponse = function(xhr, value, agent) {
    this.xhr = xhr;
	this.agent = agent;
    this.value = value || null;
    this.values = null;
    this.links_map = null;
};

jQuery.extend(AtomXmlHttpResponse.prototype, BaseHttpResponse.prototype);

AtomXmlHttpResponse.prototype.getValue = function() {
    if (!this.value) {
        this.value = this.xhr.responseXML.childNodes[0];
    }
    return this.value;
};

AtomXmlHttpResponse.prototype.getValues = function() {
    var that = this;
    
    if (!this.values) {
        var value = this.getValue();
        
        this.values = [];
        
        var entries = jQuery(this.getValue()).children('entry');
        jQuery.each(entries, function(pos, raw_entry) {
            that.values.push(new AtomXmlHttpResponse(that.xhr, raw_entry));
        });
    }
    
    return this.values;
};

AtomXmlHttpResponse.prototype.getMatchingValue = function(filter_object) {
    var value_entry_points = [];
    
    var entries = jQuery(this.getValue()).children('entry');
    var entries_length = entries.length;
    
    for (var i = 0; i < entries_length; i++) {
        var raw_entry = entries[i];
        var value = new AtomXmlHttpResponse(this.xhr, raw_entry);
        var is_match = true;
        if (typeof filter_object === 'function') {
            is_match = filter_object(value);
        } else {
            for (key in filter_object) {
                if (filter_object.hasOwnProperty(key) && jQuery(raw_entry).find(key).text() != filter_object[key]) {
                    is_match = false;
                }
            }
        }
        
        if (is_match) {
            return value;
        }
    }
    
    throw new Error('No matching value found for filter object');
};

AtomXmlHttpResponse.prototype.getLinks = function() {
    if (this.links_map) {
        return this.links_map;
    }
    
    var links_map = {};
    var links = jQuery(this.getValue()).children('link');
    jQuery.each(links, function(pos, raw_link) {
        var headers = {};
        var link = jQuery(raw_link);
        var rel = link.attr('rel');
        if (link.attr('type')) {
            headers['Content-Type'] = link.attr('type');
        }
        links_map[rel] = links_map[rel] || [];
		var absolute_href = link.attr('href');
		if (absolute_href.substr(0, 1) == '/')
		{
			absolute_href = this.agent.getBaseUrl() + absolute_href;
		}
		links_map[rel].push(new HttpAgent(absolute_href, headers));
    });
    
    this.links_map = links_map;
    return this.links_map;
};

HttpAgent.registerResponseContentTypes(['application/atom+xml'], AtomXmlHttpResponse);

XmlHttpResponse = function(xhr, value, agent) {
    this.xhr = xhr;
	this.agent = agent;
    this.value = value || null;
    this.values = null;
    this.links_map = null;
};

jQuery.extend(XmlHttpResponse.prototype, BaseHttpResponse.prototype);

XmlHttpResponse.prototype.getValue = function() {
    if (!this.value) {
        this.value = this.xhr.responseXML;
    }
    
    return this.value;
};

XmlHttpResponse.prototype.isOk = function() {
    return this.xhr.status === 200 ? true : false;
};

XmlHttpResponse.prototype.getLinks = function() {
	/* FIXME: not yet implemented */
	return {};
};

HttpAgent.registerResponseContentTypes(['text/html', 'application/xml'], XmlHttpResponse);

if (typeof define !== "undefined")
{
    define('rest-client-js', [], function () {
        return {
            "HttpAgent": HttpAgent
        };
    });
}
else if (typeof exports !== "undefined")
{
	exports.HttpAgent = HttpAgent;
}